[Core][Proposal] GetSpecifications method on Elements/Conditions

loumalouomega opened the pull request:
Fixes #5093, as discussed in #5093

loumalouomega committed:
[Core][Proposal] GetSpecifications method on Elements/Conditions

Fixes #5093, as discussed in #5093

loumalouomega committed:
Merge branch 'master' into core/GetSpecifications-method

AlejandroCornejo commented:
The @KratosMultiphysics/implementation-committee has reached consensus with this.

loumalouomega committed:
Merge branch 'master' into core/GetSpecifications-method

loumalouomega committed:
Exposing to pyhon

pooyan-dadvand commented:
What I didn't deduce in the previous discusion, why we need to pass the process info?

loumalouomega commented:
> What I didn't deduce in the previous discusion, why we need to pass the process info?

In case some additional information may be needed. It is my proposal, i can remove it

philbucher commented:
I would leave the ProcessInfo, it is used in pretty much all methods

What I don't like is adding the `Kratos::Parameters` as include, this IMO introduces quite a large dependency (and we already have quite strong dependencies)

Maybe a solution would be to separate into header/source (@loumalouomega please wait for others opinions to avoid overhead)

loumalouomega committed:
Merge branch 'master' into core/GetSpecifications-method

loumalouomega commented:
> I would leave the ProcessInfo, it is used in pretty much all methods
> 
> What I don't like is adding the `Kratos::Parameters` as include, this IMO introduces quite a large dependency (and we already have quite strong dependencies)
> 
> Maybe a solution would be to separate into header/source (@loumalouomega please wait for others opinions to avoid overhead)

I assume you mean forward declaration?, in that case you need to work with references or raw pointers (it is the only way for the compiler to assign a memory without knowing the object)

philbucher commented:
ah sure, didn't think this through...

no more complaints from me then :)

loumalouomega committed:
Merge branch 'master' into core/GetSpecifications-method

pooyan-dadvand commented:
@KratosMultiphysics/technical-committee we agree about API without passing the ProcessInfo as follow:

```c++
const Parameters GetSpecifications() const;
```

And we also discussed **without achieving the final design** about the contents of this info for the elements and conditions basing on the XML files we provide for GUI:

```Json
{
        "framework" : "eulerian",
        "symmetric_lhs": true,
        "positive_definite_lhs": false,
	"required_variables": ["DISPLACEMENT"],
	"required_dofs": ["DISPLACEMENT_X", "DISPLACEMENT_Y", "DISPLACEMENT_Z"],
        "flags_used": ["BOUNDARY", "ACTIVE"],
        "compatible_geometries": ["Tetrahedra3D4", "Hexahedra3D8"],
        "compatible_constitutive_laws": {
            "type" : ["PlaneStress","PlaneStrain"],
            "dimension" : "2D",
            "strain_size" : "3"
        }
}
```

This can be used as a basis for the discussion. Please give us your feedback to improve this proposal.

roigcarlo commented:
@rubenzorrilla suggests adding some information about the time integration in the list above. 
```json
    `is_implicit`: true,
    `element_integrates_in_time`: true
```

loumalouomega commented:
> 
> 
> @rubenzorrilla suggests adding some information about the time integration in the list above.
> 
> ```json
>     `is_implicit`: true,
>     `element_integrates_in_time`: true
> ```

I thinks as far this PR just adds the Parameters, the content can be discussed later

loumalouomega committed:
Merge branch 'master' into core/GetSpecifications-method

roigcarlo commented:
> > @rubenzorrilla suggests adding some information about the time integration in the list above.
> > ```json
> >     `is_implicit`: true,
> >     `element_integrates_in_time`: true
> > ```
> 
> I thinks as far this PR just adds the Parameters, the content can be discussed later

@KratosMultiphysics/technical-committee prefers to discuss the contents together with the API to avoid future misuses of the new method.

loumalouomega committed:
Update method (I)

loumalouomega committed:
Update method (II)

loumalouomega committed:
Merge branch 'master' into core/GetSpecifications-method

loumalouomega committed:
Merge branch 'master' into core/GetSpecifications-method

pooyan-dadvand commented:
@KratosMultiphysics/technical-committee  also asks for a test to check that all the elements registered in Kratos comply with the standard fields of this specification:
* Please propose the first version of this specification in this PR taking into account the suggestion above.
* Test should loop over all registered elements:  if the specification is not empty, check if all fields are matched with standard fields. if is empty is just bypass it to maintain the backward compatibility.

loumalouomega commented:
> 
> 
> @KratosMultiphysics/technical-committee also asks for a test to check that all the elements registered in Kratos comply with the standard fields of this specification:
> 
>     * Please propose the first version of this specification in this PR taking into account the suggestion above.

This means to add the test in this PR or just to propose the test structure? This will depend on the physical problem?. In that case may be added as a static method in the application definition. Or the previously stated is the minimum that should be considered in every problem?
> 
>     * Test should loop over all registered elements:  if the specification is not empty, check if all fields are matched with standard fields. if is empty is just bypass it to maintain the backward compatibility.

OK, see my previous comment.

loumalouomega committed:
Merge branch 'master' into core/GetSpecifications-method

pooyan-dadvand commented:
Both in this PR. The idea is to discuss the fields before merge to impose the coherency from beginning.

loumalouomega committed:
Merge branch 'master' into core/GetSpecifications-method

loumalouomega committed:
Merge branch 'master' into core/GetSpecifications-method

loumalouomega committed:
Extending documentation

loumalouomega committed:
GetSpecifications in mesh condition

loumalouomega committed:
GetSpecification in mesh element

loumalouomega committed:
Merge branch 'master' into core/GetSpecifications-method

loumalouomega committed:
Merge branch 'master' into core/GetSpecifications-method

loumalouomega committed:
Adding SpecificationsUtilities (WIP)

loumalouomega committed:
Exposing to python

loumalouomega committed:
WIP

loumalouomega committed:
Adding AddMissingVariablesFromSpecifications and finishing other methods

loumalouomega committed:
Adding AddMissingDofsFromSpecifications

loumalouomega committed:
Fix

loumalouomega committed:
Merge branch 'master' into core/GetSpecifications-method

loumalouomega committed:
Adding new methods (WIP)

loumalouomega committed:
Exposing to python

loumalouomega committed:
Completing flags methods

loumalouomega committed:
Correction

loumalouomega committed:
Framework method

loumalouomega committed:
Fix

loumalouomega committed:
Implement for symmetric

loumalouomega committed:
Define DeterminePositiveDefiniteLHS

loumalouomega committed:
Define DetermineIfImplicitSimulation

loumalouomega committed:
DetermineIfRequiresTimeIntegration

loumalouomega committed:
Merge branch 'master' into core/GetSpecifications-method

loumalouomega committed:
Replace tabulation

loumalouomega committed:
Merge branch 'master' into core/GetSpecifications-method

loumalouomega committed:
Merge branch 'master' into core/GetSpecifications-method

loumalouomega committed:
Merge branch 'master' into core/GetSpecifications-method

loumalouomega committed:
Merge branch 'master' into core/GetSpecifications-method

loumalouomega committed:
Merge branch 'core/GetSpecifications-method' of https://github.com/KratosMultiphysics/Kratos into core/GetSpecifications-method

loumalouomega committed:
Adding to base solid element

loumalouomega committed:
Update

loumalouomega committed:
[Core] Missing dummy elements

Triangle2D6, Quadrilateral2D8, Quadrilateral2D9, Prism3D15, Hexahedra3D20 and Hexahedra3D27 were missing

loumalouomega committed:
Adding map string geometry type

loumalouomega committed:
Complete

loumalouomega committed:
Actual implementation

loumalouomega committed:
Adding more maps

loumalouomega committed:
New sepecifications

loumalouomega committed:
Defining

loumalouomega committed:
Merge branch 'master' into core/GetSpecifications-method

loumalouomega committed:
Merge branch 'master' into core/GetSpecifications-method

adityaghantasala commented:
On behalf of @KratosMultiphysics/implementation-committee  

We were thinking it would be benificial to have some empty set of keys in the base element/condition. Current implementation passes  empty Parameters. But having a set of consentuated set of keys like `required_variables` and `require_dofs` (every element/condition requires dofs and variables) etc would encourage developers to think of kratos element/condition in a more specific way.  Having an empty parameters object there may encourage misuse of this nice functionality. 

@loumalouomega  could you already propose the first version of this set ?

loumalouomega committed:
Minimal parameters (I)

loumalouomega committed:
Minimal parameters (II)

loumalouomega commented:
> @loumalouomega could you already propose the first version of this set ?

I need to implement the test. but it is almost ready

loumalouomega committed:
Merge branch 'master' into core/GetSpecifications-method

loumalouomega committed:
Merge branch 'master' into core/GetSpecifications-method

loumalouomega committed:
Merge branch 'master' into core/GetSpecifications-method

loumalouomega committed:
Moving definition 

(wrong files)

loumalouomega committed:
Revert

loumalouomega committed:
Conflict

loumalouomega committed:
Adding to BaseLoadCondition

loumalouomega committed:
WIP

loumalouomega committed:
Register tests

loumalouomega committed:
Correction

loumalouomega committed:
Fix

loumalouomega committed:
Activate test

loumalouomega committed:
New test

loumalouomega committed:
Test for DetermineFramework

loumalouomega committed:
Test for DetermineSymmetricLHS

loumalouomega committed:
Test for DeterminePositiveDefiniteLHS

loumalouomega committed:
Correct DetermineIfCompatibleGeometries

loumalouomega committed:
Test for DetermineIfCompatibleGeometries

loumalouomega committed:
Test for DetermineIfImplicitSimulation

loumalouomega committed:
Test for DetermineIfRequiresTimeIntegration

loumalouomega committed:
Test for CheckCompatibleConstitutiveLaws

loumalouomega commented:
@KratosMultiphysics/implementation-committee Tests have been added

loumalouomega committed:
Merge branch 'master' into core/GetSpecifications-method

loumalouomega committed:
Extend test

loumalouomega committed:
Minor

loumalouomega committed:
Using pointer instead of list

loumalouomega committed:
Activate test

loumalouomega committed:
Triggering a failing test

loumalouomega committed:
Merge branch 'master' into core/GetSpecifications-method

loumalouomega committed:
Minor fix

loumalouomega committed:
Merge branch 'master' into core/GetSpecifications-method

loumalouomega committed:
Merge branch 'master' into core/GetSpecifications-method

loumalouomega committed:
Merge branch 'master' into core/GetSpecifications-method

loumalouomega committed:
Merge branch 'master' into core/GetSpecifications-method

loumalouomega committed:
Merge branch 'master' into core/GetSpecifications-method

loumalouomega committed:
Update condition name to SurfaceCondition3D3N

RiccardoRossi commented:
some minor comments after reviewing in @KratosMultiphysics/technical-committee :

1 - is_implicit -> time_integration : [ ... here list possible compatible types ... "implicit", "explicit", "static"]
2 - output --> missing to list potential outputs. so 
    - output on Gauss Points
    - nodal historical output
    - nodal non historical output
    - entity output (the result of GetValue on the element)
3 - we should define a map geometry_type/name
4 - more filter features for CL, in particular deformation measure and stress measure required
5 - "Documentation" add a free field where one can point to the documentation and maybe a reference to the formulation

The idea is to use this as a mechanism for generating automatic docs that go along with the code. For this to work we will need to use this extensively in the testing (which will have to be done app by app to take into account the specificities of each)

loumalouomega committed:
Merge branch 'master' into core/GetSpecifications-method

loumalouomega commented:
> some minor comments after reviewing in @KratosMultiphysics/technical-committee :
> 
> 1 - is_implicit -> time_integration : [ ... here list possible compatible types ... "implicit", "explicit", "static"]

Static is implicit... (I think we are mixing time integration with resolution type)

> 2 - output --> missing to list potential outputs. so
> - output on Gauss Points
> - nodal historical output
> - nodal non historical output
> - entity output (the result of GetValue on the element)

OK

> 3 - we should define a map geometry_type/name

??¿¿

> 4 - more filter features for CL, in particular deformation measure and stress measure required

I require a defined list in order to implement it correctly

> 5 - "Documentation" add a free field where one can point to the documentation and maybe a reference to the formulation

OK

> 
> The idea is to use this as a mechanism for generating automatic docs that go along with the code. For this to work we will need to use this extensively in the testing (which will have to be done app by app to take into account the specificities of each)

OK

loumalouomega committed:
Merge branch 'master' into core/GetSpecifications-method

loumalouomega committed:
Adding time_integration

loumalouomega committed:
Enums

loumalouomega committed:
Define DetermineTimeIntegration

loumalouomega committed:
Adding test

loumalouomega committed:
Defining output and documentation

loumalouomega committed:
Adding sort

loumalouomega committed:
Add GetDocumentation method

loumalouomega committed:
Adding method to mesh entities

loumalouomega committed:
Adding implementation

loumalouomega committed:
Adding documentation test

rubenzorrilla commented on kratos/utilities/specifications_utilities.h:
```diff
@@ -52,6 +52,26 @@ namespace Kratos
+    enum class Framework
+    {
+        Lagrangian = 0,
+        Euler      = 1,
```
```suggestion
        Eulerian      = 1,
```

loumalouomega replied:
BTW, for now this is not used. Currently I use the string, this is just an idea to later remove if the @KratosMultiphysics/technical-committee says so. PS: enums are a pain in python

rubenzorrilla commented on kratos/utilities/specifications_utilities.h:
```diff
@@ -52,6 +52,26 @@ namespace Kratos
+        Euler      = 1,
+        ALE        = 2
+    };
+
```
Should we add the Updated Lagrangian to the list too?

loumalouomega replied:
Nope, Updated Lagrangian is still Lagrangian. The point of Lagrangian is communicate to the solver that the nodes of these elements/conditions require to be moved after solving the step

loumalouomega replied:
Well, fort now this just tells the framework, for now it is not taken into account for nothing

rubenzorrilla replied:
Your argument makes sense to me (the updated is still Lagrangian).

rubenzorrilla commented on kratos/utilities/specifications_utilities.cpp:
```diff
@@ -442,15 +552,15 @@ std::string DetermineFramework(ModelPart& rModelPart)
             } else {
                 CompareElementsAndConditionsUtility::GetRegisteredName(*it_cond_begin, condition_name);
-                KRATOS_WARNING("SpecificationsUtilities") << "Framework:" << r_framework << " is used by the condition: " << condition_name << ". But also the framework: " << framework << " is used" << std::endl; 
+                KRATOS_WARNING("SpecificationsUtilities") << "Framework:" << r_framework << " is used by the condition: " << condition_name << ". But also the framework: " << framework << " is used" << std::endl;
```
Not sure if this has to be a warning. It is perfectly possible to mix Eulerian elements with ALE ones.

loumalouomega replied:
And Lagrangian with Eulerian, or ALE, I agree. This is a draft, I am doing what the @KratosMultiphysics/technical-committee  tells me, and the lack of details makes me to take some ambiguous decisions

philbucher replied:
just to say that we are also not sure what should be there...

loumalouomega replied:
> just to say that we are also not sure what should be there...

![imagen](https://user-images.githubusercontent.com/19991680/78149957-ea871a00-7436-11ea-955c-1740e8e57e48.png)

loumalouomega committed:
Euler->Eulerian

Co-Authored-By: Rubén Zorrilla <rubenzorrillamartinez@hotmail.com>

loumalouomega committed:
Merge master

loumalouomega committed:
Merge branch 'master' into core/GetSpecifications-method

loumalouomega committed:
Merge branch 'master' into core/GetSpecifications-method

loumalouomega committed:
Merge branch 'master' into core/GetSpecifications-method

loumalouomega committed:
Merge branch 'master' into core/GetSpecifications-method

pooyan-dadvand commented on applications/StructuralMechanicsApplication/custom_conditions/base_load_condition.cpp:
```diff
@@ -366,6 +366,38 @@ void BaseLoadCondition::AddExplicitContribution(
+        "required_dofs"              : ["DISPLACEMENT_X","DISPLACEMENT_Y","DISPLACEMENT_Z"],
+        "flags_used"                 : [],
+        "compatible_geometries"      : ["Line2D2", "Line2D3", "Triangle3D3", "Triangle3D6", "Quadrilateral3D4", "Quadrilateral3D8", "Quadrilateral3D9"],
+        "is_implicit"                : true,
```
It is redundant as it is given in the "time_integration" field.

loumalouomega replied:
I just follow the requirements from the @KratosMultiphysics/technical-committee, if the @KratosMultiphysics/technical-committee agrees I will remove it

pooyan-dadvand replied:
@KratosMultiphysics/technical-committee agrees on removing it and check the `is_implicit `by `[time_integration].Has("implicit")`

loumalouomega committed:
Merge branch 'master' into core/GetSpecifications-method

loumalouomega committed:
Removing is_implicit

loumalouomega committed:
Removing from derived classses

loumalouomega committed:
Removing from structural

loumalouomega committed:
Update utilities

loumalouomega committed:
Update export

loumalouomega committed:
Update tests

loumalouomega committed:
Merge branch 'master' into core/GetSpecifications-method

Commit:
Remove legacy code

Commit:
Legacy variable type

Commit:
Merge branch 'master' into core/GetSpecifications-method

loumalouomega committed:
Merge branch 'master' into core/GetSpecifications-method

loumalouomega committed:
Minor update commend

loumalouomega committed:
Udpate method

loumalouomega committed:
Update utilities

loumalouomega committed:
Merge branch 'master' into core/GetSpecifications-method

loumalouomega committed:
Merge branch 'master' into core/GetSpecifications-method

loumalouomega committed:
Adding CheckGeometricalPolynomialDegree

loumalouomega committed:
Expose CheckGeometricalPolynomialDegree

loumalouomega committed:
Defining required_polynomial_degree_of_geometry in element

loumalouomega committed:
Adding required_polynomial_degree_of_geometry in condition

loumalouomega committed:
Adding required_polynomial_degree_of_geometry to mesh element

loumalouomega committed:
Adding required_polynomial_degree_of_geometry to mesh condition

loumalouomega committed:
Adding required_polynomial_degree_of_geometry to solid element

loumalouomega committed:
Adding required_polynomial_degree_of_geometry to load condition

loumalouomega committed:
Extend test with required_polynomial_degree_of_geometry

loumalouomega committed:
Merge branch 'master' into core/GetSpecifications-method

pooyan-dadvand reviewed (approved):
Thanks @loumalouomega for the effort!
